### PR TITLE
std.mem.bytesAsValue for byte-aligned packed structs

### PR DESCRIPTION
**Pull request by motiejus**
Currently `bytesAsValue` works with only word-aligned structs; technically, when `@sizeOf(T) == @divExact(@bitSizeOf(T), 8)`.

This diff enables converting bytes to non-word-aligned struct pointers. Though still requires them to be byte-aligned.

`asBytes` is not so easy; I will file a separate task for it.

### CODE REVIEW

**Comment by motiejus** [comment]
re. drone-ci failures: this command seems to have worked for me:

```
$ ./build/zig build -Dtarget=arm-linux-musleabihf
<...>
$ file zig-out/bin/zig
zig-out/bin/zig: ELF 32-bit LSB executable, ARM, EABI5 version 1 (SYSV), statically linked, with debug_info, not stripped
$ 
```

**Comment by Vexu** [comment]
The test you added passes without changing the `@sizeOf`, do you have a test case that doesn't currently pass?

> re. drone-ci failures: this command seems to have worked for me:

It is caused by a broken commit in master, don't know why the CI is using master instead of your branch.

**Comment by motiejus** [comment]
> The test you added passes without changing the `@sizeOf`, do you have a test case that doesn't currently pass?

This is a mixup, thanks for noticing. Will do so likely tomorrow.

**Comment by motiejus** [comment]
> > The test you added passes without changing the `@sizeOf`, do you have a test case that doesn't currently pass?
> 
> This is a mixup, thanks for noticing. Will do so likely tomorrow.

Updated. It wasn't easy to find a packed struct that satisfies the assertion; I would love to learn more about ([in `@sizeOf` docs](https://ziglang.org/documentation/master/#sizeOf)) though:

```
This size may contain padding bytes. 
```

**Comment by motiejus** [comment]
Changing `asBytes` to assume byte-aligned data fails to compile stage1, as there are some non-byte-aligned structs. So `asBytes` cannot be changed as-is. I also spent an hour or so mirroring the change to `asBytes`, to no avail.

I will leave this PR to `bytesAsValue` alone and file an issue for `asBytes` to discuss it further.

File: lib/std/mem.zig [hunk 1]
```diff
@@ -2758,6 +2758,13 @@ test "bytesAsValue preserves pointer attributes" {
     try testing.expectEqual(in.alignment, out.alignment);
 }
 
+test "bytesAsValue works with non-word-aligned structs" {
```
Comment (andrewrk) [inline lib/std/mem.zig:2761]: Please use a different word than "word" which is ambiguous.

File: lib/std/mem.zig [hunk 2]
```diff
@@ -2758,6 +2758,13 @@ test "bytesAsValue preserves pointer attributes" {
...
+    const padded = packed struct { a: u14, b: u10 };
+    // for test to be meaningful @sizeOf must think the struct is padded.
+    assert(@sizeOf(padded) != @divExact(@bitSizeOf(padded), 8));
```
Comment (andrewrk) [inline lib/std/mem.zig:2764]: This assertion trips on latest master:

```
$ ./stage1/bin/zig test test.zig 
Test [1/1] test "example"... thread 2925231 panic: reached unreachable code
/home/andy/Downloads/zig/lib/std/debug.zig:260:14: 0x206aab in std.debug.assert (test)
    if (!ok) unreachable; // assertion failure
             ^
/home/andy/Downloads/zig/build-release/test.zig:7:11: 0x2076de in test "example" (test)
    assert(@sizeOf(padded) != @divExact(@bitSizeOf(padded), 8));
          ^
```

File: lib/std/mem.zig [hunk 3]
```diff
@@ -2758,6 +2758,13 @@ test "bytesAsValue preserves pointer attributes" {
...
+    _ = bytesAsValue(padded, &[3]u8{ 0, 0, 0 });
```
Comment (andrewrk) [inline lib/std/mem.zig:2765]: The test should write through the pointer and then observe an expected outcome

**Review by andrewrk — CHANGES_REQUESTED** [review]

**Comment by motiejus** [comment]
I think I no longer want to do this, because:
- we cannot make a symmetric change to `asBytes` (most importantly).
- this is only for stage1 (at the time I did not test stage2).
- I am adding the assertion to my packed structs: `@sizeOf(S) * 8 == @bitSizeOf(S)` and life is easy.

**Pull request closed without merging**